File: submissions/06-gemini/wasm/verify.mjs
import { readFile } from "node:fs/promises";

const wasmPath = process.argv[2] ?? "gemini.wasm";
const bytes = await readFile(wasmPath);
const { instance } = await WebAssembly.instantiate(bytes, {});

const checks = [
  ["add(2,3)", instance.exports.add(2, 3), 5],
  ["gemini_sum(12,25)", instance.exports.gemini_sum(12, 25), 37],
  ["gemini_pack_size()", instance.exports.gemini_pack_size(), 37],
  ["gemini_skill_check(6)", instance.exports.gemini_skill_check(6), 36],
  ["gemini_gif_size()", instance.exports.gemini_gif_size(), 17822],
];

for (const [label, got, want] of checks) {
  if (got !== want) {
    throw new Error(`${label}: got ${got}, want ${want}`);
  }
  console.log(`${label} = ${got} PASS`);
}

const gifPtr = instance.exports.gemini_gif_ptr();
console.log(`gemini_gif_ptr() = ${gifPtr} (WASM offset) PASS`);

console.log(`${wasmPath}: ${bytes.length} bytes, zero-import module verified`);

File: submissions/02-esp32-oracle/wasm/verify.py
#!/usr/bin/env python3
"""Verify the Sentinel wasm exports with wasmtime."""

from __future__ import annotations

import sys
from pathlib import Path

from wasmtime import Instance, Module, Store


def main() -> int:
    wasm_path = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("sentinel.wasm")
    wasm = wasm_path.read_bytes()

    store = Store()
    module = Module(store.engine, wasm)
    instance = Instance(store, module, [])
    exports = instance.exports(store)

    flame = exports["flame"]
    sense = exports["sense"]

    checks = [
        ("flame(10)", flame(store, 10), 385),
        ("sense(3,4)", sense(store, 3, 4), 25),
    ]

    for label, got, want in checks:
        if got != want:
            raise SystemExit(f"{label}: got {got}, want {want}")
        print(f"{label} = {got} PASS")

    print(f"{wasm_path.name}: {len(wasm)} bytes, exports verified")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
